### PR TITLE
Lazy sympy

### PR DESCRIPTION
**Pull request by cscherrer**
Because the `__init__` for all of Soss had been in `symbolic.jl`, I had to shuffle things around a little bit. So now, ...

- `__init__` is in `Soss.jl`, and checks for the presence of `SymPy`
- If it's there, it `include`s `_SymPy.jl`
- `_SymPy.jl` just wraps `symbolic.jl` and `codegen.jl` in a module

### CODE REVIEW

**Comment by cscherrer** [comment]
Oh, I just realized...

- Module hierarchy should be reflected in folder hierarchy
- The `README.jmd` needs to be updated and re-built to account for the `SymPy` module
- Is `SymPy` a bad name for a module? ...Kind of? But I like the idea of allowing for other ways of symbolic simplification

File: Project.toml [hunk 1]
```diff
@@ -32,7 +33,6 @@ SimplePosets = "b2aef97b-4721-5af9-b440-0bad754dc5ba"
 SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 StatsFuns = "4c63d2b9-4356-54db-8cca-17b64c39e42c"
-SymPy = "24249f21-da20-56a4-8eb1-6a02cf4ae2e6"
```
Comment (baggepinnen) [inline Project.toml:35]: It's probably a good idea to keep SymPy under the [extras] section for testing purposes. Here's how the project file  looks for MCM, all the packages listed under extras and targets/test can be used while testing

```
[extras]
ControlSystems = "a6e380b2-a6ca-5380-bf3e-84a91bcd477e"
ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
Measurements = "eff96d63-e80a-5855-80a2-b1b0885c5ab7"
Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
SLEEFPirates = "476501e8-09a2-5ece-8869-fb82de89a1fa"
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[targets]
test = ["ControlSystems", "ForwardDiff", "Measurements", "SLEEFPirates", "Test", "Plots"]
```
They will only be installed if tests are run, not when installing the package

Comment (cscherrer) [reply Project.toml:35]: Great point, thanks. As I understand, this has changed recently, and the "modern" approach is to have separate `Project.toml`s for the project itself, `docs`, and `test`. I'll add SymPy to the... I want to say "latter two", is that a thing? :)

**Comment by cscherrer** [comment]
Looks like I had misunderstood something about `Requires.jl`. The optional package still needs to be in the `Project.toml` dependencies, but it's not loaded using `using` it. So for my machine, I started out at 97 seconds, then down to 30, and now...
```julia
julia> @time using Soss
 13.123677 seconds (28.75 M allocations: 1.549 GiB, 5.14% gc time)
```

SymPy isn't working yet though, so it crashes when `using` it

File: Project.toml (hunk 1 above)
Comment (baggepinnen) [reply Project.toml]: Using a separate project file for docs is common. I haven't seen it for tests before but that might work as well, although I can't find any info about it at https://julialang.github.io/Pkg.jl/v1/toml-files/
In either case, not automatically installing sympy when Soss is installed would be great

**Comment by baggepinnen** [comment]
I think you have to do some workaround like `using .SymPy` in the `@require`d file, notice the `.`.

**Comment by cscherrer** [comment]
I think I see what you're getting at, but with the dot it crashes sooner. Currently I'm getting this warning:
```julia
julia> using Soss, SymPy
┌ Warning: Error requiring SymPy from Soss:
│ LoadError: LoadError: TypeError: in Type{...} expression, expected UnionAll, got typeof(Soss.SymbolicSymPy.For)
│ Stacktrace:
│  [1] top-level scope at /home/chad/git/Soss.jl/src/symbolic/symbolic.jl:307
│  [2] include(::Module, ::String) at ./Base.jl:377
│  [3] include(::String) at /home/chad/git/Soss.jl/src/_sympy.jl:1
│  [4] top-level scope at /home/chad/git/Soss.jl/src/_sympy.jl:2
│  [5] include(::Module, ::String) at ./Base.jl:377
│  [6] include(::String) at /home/chad/git/Soss.jl/src/Soss.jl:1
│  [7] top-level scope at /home/chad/git/Soss.jl/src/Soss.jl:54
│  [8] eval at ./boot.jl:331 [inlined]
│  [9] eval at /home/chad/git/Soss.jl/src/Soss.jl:1 [inlined]
│  [10] (::Soss.var"#148#151")() at /home/chad/.julia/packages/Requires/qy6zC/src/require.jl:85
│  [11] err(::Any, ::Module, ::String) at /home/chad/.julia/packages/Requires/qy6zC/src/require.jl:42
│  [12] (::Soss.var"#147#150")() at /home/chad/.julia/packages/Requires/qy6zC/src/require.jl:84
│  [13] withpath(::Any, ::String) at /home/chad/.julia/packages/Requires/qy6zC/src/require.jl:32
│  [14] (::Soss.var"#146#149")() at /home/chad/.julia/packages/Requires/qy6zC/src/require.jl:83
│  [15] #invokelatest#1 at ./essentials.jl:712 [inlined]
│  [16] invokelatest at ./essentials.jl:711 [inlined]
│  [17] foreach at ./abstractarray.jl:1919 [inlined]
│  [18] loadpkg(::Base.PkgId) at /home/chad/.julia/packages/Requires/qy6zC/src/require.jl:22
│  [19] #invokelatest#1 at ./essentials.jl:712 [inlined]
│  [20] invokelatest at ./essentials.jl:711 [inlined]
│  [21] require(::Base.PkgId) at ./loading.jl:930
│  [22] require(::Module, ::Symbol) at ./loading.jl:922
│  [23] eval(::Module, ::Any) at ./boot.jl:331
│  [24] eval_user_input(::Any, ::REPL.REPLBackend) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
│  [25] run_backend(::REPL.REPLBackend) at /home/chad/.julia/packages/Revise/xeiUU/src/Revise.jl:1070
│  [26] (::Revise.var"#85#87"{REPL.REPLBackend})() at ./task.jl:358
│ in expression starting at /home/chad/git/Soss.jl/src/symbolic/symbolic.jl:307
│ in expression starting at /home/chad/git/Soss.jl/src/_sympy.jl:2
```
But then I get weird behavior after that. If I do `]test` it gives me this warning and then crashes shortly after. Namespace weirdness :(

File: Project.toml (hunk 1 above)
Comment (sethaxen) [reply Project.toml]: Yeah, I haven't seen it for tests either. I don't think it would be necessary due to `extras`.

**Comment by sethaxen** [comment]
Why is the code that uses SymPy in its module? It shouldn't be necessary for Requires, and not having it in its own module might help with the namespace weirdness you're seeing.

Also, 13s! Even better than in https://github.com/cscherrer/Soss.jl/issues/116.

File: docs/Project.toml [hunk 2]
```diff
@@ -1,3 +1,4 @@
 [deps]
 Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
 Soss = "8ce77f84-9b61-11e8-39ff-d17a774bf41c"
+SymPy = "24249f21-da20-56a4-8eb1-6a02cf4ae2e6"
```
Comment (sethaxen) [inline docs/Project.toml:4]: Should probably also add the `compat` entry here too.

File: Project.toml [hunk 3]
```diff
@@ -24,6 +24,7 @@ PyCall = "438e738f-606a-5dbb-bf0a-cddfbfd45ab0"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
+Requires = "ae029012-a4dd-5104-9daa-d747884805df"
```
Comment (sethaxen) [inline Project.toml:27]: Would be good to have a `compat` for Requires. Should be able to support both versions 0.5 and 1.

**Comment by cscherrer** [comment]
> Why is the code that uses SymPy in its module?

Maybe I'm misunderstanding something here. My mental model is that `include` inlines code from other files, so Julia doesn't "know" about the file or directory structure in any way. 

Is there a standard way to set up a file hierarchy to match the module hierarchy?

File: docs/Project.toml (hunk 2 above)
Comment (cscherrer) [reply docs/Project.toml:4]: Oh, hadn't thought of that. Should there be separate `compat` entries for docs and test? I'd think those fall back to whatever is given for the overall project

**Comment by DilumAluthge** [comment]
@cscherrer I'm going to close this in favor of https://github.com/cscherrer/Soss.jl/issues/170

**Pull request closed without merging**